pkg/ucp/initializer/service.go

```diff
@@ -0,0 +1,135 @@
+	"github.com/radius-project/radius/pkg/ucp/hosting"
+
+	//ucpoptions "github.com/radius-project/radius/pkg/ucp/hostoptions"
+	"github.com/radius-project/radius/pkg/ucp/ucplog"
+)
+
+// Service implements the hosting.Service interface for registering manifests.
+type Service struct {
+	options *ucp.Options
+}
+
+var _ hosting.Service = (*Service)(nil)
+
+// NewService creates a server to register manifests.
+func NewService(options *ucp.Options) *Service {
+	return &Service{
+		options: options,
+	}
+}
+
+// Name gets this service name.
+func (s *Service) Name() string {
+	return "initializer"
+}
+
+func waitForServer(ctx context.Context, host, port string, retryInterval time.Duration, connectionTimeout time.Duration, timeout time.Duration) error {
+	address := net.JoinHostPort(host, port)
+	deadline := time.Now().Add(timeout)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("connection attempts canceled or timed out: %w", ctx.Err())
+		default:
+			conn, err := net.DialTimeout("tcp", address, connectionTimeout)
+			if err == nil {
+				conn.Close()
+				return nil
+			}
+
+			if time.Now().After(deadline) {
+				return fmt.Errorf("failed to connect to %s after %v: %w", address, timeout, err)
+			}
+
+			time.Sleep(retryInterval)
+		}
+	}
+}
+
+func (w *Service) Run(ctx context.Context) error {
+	logger := ucplog.FromContextOrDiscard(ctx)
+
+	if w.options.UCP == nil || w.options.UCP.Endpoint() == "" {
+		return fmt.Errorf("connection to UCP is not set")
+	}
+
+	// Parse the endpoint URL and extract host and port
+	parsedURL, err := url.Parse(w.options.UCP.Endpoint())
+	if err != nil {
+		return fmt.Errorf("failed to parse endpoint URL: %w", err)
+	}
+
+	hostName, port, err := net.SplitHostPort(parsedURL.Host)
+	if err != nil {
+		return fmt.Errorf("failed to split host and port: %w", err)
+	}
+	logger.Info("Parsed Host and Port", "host", hostName, "port", port)
+
+	// Attempt to connect to the server
+	err = waitForServer(ctx, hostName, port, 500*time.Millisecond, 500*time.Millisecond, 5*time.Second)
+	if err != nil {
+		logger.Error(err, "Server is not available for manifest registration")
+		return nil
+	}
+
+	// Server is up, proceed to register manifests
+	manifestDir := w.options.Config.Initialization.ManifestDirectory
+	if manifestDir == "" {
+		logger.Info("No manifest directory specified")
+		return nil
+	}
+
+	if _, err := os.Stat(manifestDir); os.IsNotExist(err) {
+		return fmt.Errorf("manifest directory does not exist: %w", err)
+	} else if err != nil {
+		return fmt.Errorf("error checking manifest directory: %w", err)
+	}
+
+	clientOptions := sdk.NewClientOptions(w.options.UCP)
+
+	clientFactory, err := v20231001preview.NewClientFactory(&aztoken.AnonymousCredential{}, clientOptions)
+	if err != nil {
+		return fmt.Errorf("error creating client factory: %w", err)
+	}
+
+	// Proceed with registering manifests
+	if err := manifest.RegisterDirectory(ctx, clientFactory, "local", manifestDir, nil); err != nil {
```

Comment:
Do we want to pass in the logger here? It might be helpful to see each individual thing getting logged. If it's too much spam we can set it to debug level.

Reply:
updated